plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs

```diff
@@ -82,53 +117,76 @@ computeCek env con var@(Var _ name)       = case lookupName name env of
-applyEvaluate :: Environment -> Environment -> Context -> Plain Value -> Plain Value -> EvaluationResult
-applyEvaluate funEnv argEnv con (LamAbs _ name _ body) arg =
-    computeCek (extendEnvironment name arg argEnv funEnv) con body
-applyEvaluate funEnv _      con fun                    arg =
+applyEvaluate
+    :: VarEnv -> VarEnv -> Context -> Plain Value -> Plain Value -> CekM EvaluationResult
+applyEvaluate funVarEnv argVarEnv con (LamAbs _ name _ body) arg =
+    withVarEnv (extendVarEnv name arg argVarEnv funVarEnv) $ computeCek con body
+applyEvaluate funVarEnv _         con fun                    arg =
     let term = Apply () fun arg in
         case termAsPrimIterApp term of
             Nothing                       ->
-                throw $ MachineException NonPrimitiveApplicationMachineError term
-            Just (IterApp headName spine) ->
-                case runQuote $ applyBuiltinName headName spine of
-                    ConstAppSuccess term' -> returnCek funEnv con term'
-                    ConstAppFailure       -> EvaluationFailure
-                    ConstAppStuck         -> returnCek funEnv con term
-                    ConstAppError err     -> throw $ MachineException (ConstAppMachineError err) term
-
--- | Evaluate a term using the CEK machine. May throw a 'MachineException'.
-evaluateCek :: Term TyName Name () -> EvaluationResult
-evaluateCek = computeCek (Environment IntMap.empty) []
-
--- | Run a program using the CEK machine. May throw a 'MachineException'.
--- Calls 'evaluateCek' under the hood, so the same caveats apply.
-runCek :: Program TyName Name () -> EvaluationResult
-runCek (Program _ _ term) = evaluateCek term
+                throwError $ MachineException NonPrimitiveApplicationMachineError term
+            Just (IterApp headName spine) -> do
+                constAppResult <- runQuote <$> applyStagedBuiltinName headName spine
+                withVarEnv funVarEnv $ case constAppResult of
+                    ConstAppSuccess res -> returnCek con res
+                    ConstAppFailure     -> pure EvaluationFailure
+                    ConstAppStuck       -> returnCek con term
+                    ConstAppError   err ->
+                        throwError $ MachineException (ConstAppMachineError err) term
+
+-- | Apply a 'StagedBuiltinName' to a list of 'Value's.
+applyStagedBuiltinName :: StagedBuiltinName -> [Plain Value] -> CekM (Quote ConstAppResult)
+applyStagedBuiltinName (DynamicStagedBuiltinName name) args = do
+    mayMean <- lookupDynamicBuiltinName name
+    pure $ case mayMean of
+        -- returns 'ConstAppFailure' in case a dynamic built-in is out of scope.
+        Nothing                                -> pure ConstAppFailure
+        Just (DynamicBuiltinNameMeaning sch x) -> applyTypeSchemed sch x args
+applyStagedBuiltinName (StaticStagedBuiltinName  name) args = pure $ applyBuiltinName name args
+
+-- | Evaluate a term using the CEK machine.
+evaluateCekCatch
```

Comment:
Hm, I think I'd prefer this to be the main interface, but that's a separate discussion.